feat(api, dashboard): introduces step content generation through LLMs fixes NV-6931

merrcury opened the pull request:
### What changed? Why was the change needed?
This implements the step content generation through LLMs. 

### Screenshots

<!-- If the changes are visual, include screenshots or screencasts. -->

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Related enterprise PR

<!-- A link to a dependent pull request  -->

### Special notes for your reviewer

<!-- Specific instructions or considerations you want to highlight for the reviewer. -->

</details>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added AI-powered content generation for notification templates across email, SMS, push, in-app, and chat channels
  * Introduced "Compose with AI" functionality in the workflow editor
  * Added interactive AI suggestions dialog with prompt input and content preview
  * Enabled variable substitution and payload suggestions for generated content

* **Dependencies**
  * Added OpenAI and AI SDK integrations for content generation capabilities

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

merrcury committed:
Merge branch 'next' of github.com:novuhq/novu into feat/step-creation-with-llm

merrcury committed:
feat(api): integrate AI content generation with new AI module and endpoints

- Added AiModule with AiController for handling AI content generation requests.
- Implemented GenerateContentUseCase to interact with OpenAI and generate content based on user prompts.
- Introduced new DTOs for request and response handling in AI content generation.
- Updated pnpm-lock.yaml and package.json to include necessary AI SDK dependencies.
- Enhanced dashboard with AI suggestions dialog for content generation.
- Added hooks for managing AI content generation in the dashboard.

This commit establishes a foundation for AI-driven content generation across the application.

merrcury committed:
feat(dashboard): enhance AI suggestions dialog with dynamic prompts and loading/error states

- Introduced a new set of prompt suggestions categorized by step type for better user experience.
- Added functionality to load more suggestions dynamically.
- Implemented custom loading and error states for email previews, improving visual feedback during content generation.
- Added shimmer animation for loading indicators to enhance UI responsiveness.
- Updated CSS to support new animations and styles for loading states.

merrcury committed:
feat(dashboard): implement AI step generation feature flag and enhance AI button UI

- Added a feature flag for AI step generation to control the visibility of AI-related functionalities.
- Updated the AI button to only display when the feature flag is enabled, improving user experience.
- Enhanced button styling and animations for better visual appeal and interaction feedback.
- Introduced new gradient backgrounds and icon updates for the AI button to align with design standards.

merrcury committed:
feat(dashboard): enhance AI step editor with schema integration and payload handling

- Added a helper function to infer JSON Schema types from values, improving payload management.
- Integrated schema property management to dynamically add new properties based on AI-generated payloads.
- Updated the AI insert functionality to handle payload schema updates when enabled, enhancing the editor's capabilities.
- Removed unnecessary console logs to clean up the codebase.

merrcury committed:
refactor(dashboard): AI suggestions dialog and introduce new components

- Consolidated AI suggestions logic by creating dedicated components for initial, loading, and error states.
- Introduced a constants file for prompt suggestions categorized by step type, enhancing maintainability.
- Added utility functions for parsing variables and managing nested values, improving code clarity.
- Implemented a content preview component to render AI-generated content dynamically based on step type.
- Enhanced the overall structure of the AI suggestions dialog for better readability and performance.

merrcury committed:
chore(dashboard): update dependencies for AI SDK and related packages

- Downgraded @ai-sdk/react from 2.1.0-beta.28 to 2.0.108 for compatibility.
- Downgraded ai from 5.1.0-beta.28 to 5.0.108 to align with the new SDK version.
- Updated pnpm-lock.yaml to reflect changes in dependency versions and their nested dependencies.
- Ensured all related packages are consistent with the new versions to maintain stability.

merrcury committed:
chore(api): downgrade AI SDK dependencies for compatibility

- Downgraded @ai-sdk/openai from 2.1.0-beta.16 to 2.0.79 to align with the new versioning strategy.
- Downgraded ai from 5.1.0-beta.28 to 5.0.108 to maintain consistency across dependencies.
- Updated pnpm-lock.yaml to reflect the changes in dependency versions and their resolutions.
- Removed deprecated beta versions to streamline the dependency tree and enhance stability.

merrcury committed:
feat(dashboard): enhance InboxPreview component with gradient control

- Updated InboxPreview component to include a showGradient prop, allowing control over gradient display in various states (content preview, error state, loading state).
- Adjusted related components to pass the new prop, ensuring consistent behavior across different use cases.

merrcury committed:
refactor(api): replace Logger with PinoLogger in GenerateContentUseCase

- Updated GenerateContentUseCase to utilize PinoLogger for improved logging capabilities.
- Changed log statements from `this.logger.log` to `this.logger.info` for better log level semantics.
- Set logger context in the constructor to enhance log clarity and traceability.

merrcury committed:
chore(api): update lint command to include diagnostic level

- Modified the lint command in package.json to set the diagnostic level to error, enhancing linting feedback and ensuring stricter code quality checks.

merrcury committed:
feat(dashboard): enhance ContentPreview component layout and add user guidance

- Updated the layout of the ContentPreview component to improve visual structure and responsiveness.
- Added a user guidance message to inform users that they can edit content after insertion into the step editor, enhancing usability.

merrcury committed:
feat(dashboard): enhance AI suggestions dialog with existing content handling

- Added a new prop to the AiSuggestionsDialog component to check for existing content, improving user feedback during content insertion.
- Updated the footer buttons to conditionally display "Replace content" or "Insert content" based on the presence of existing content, enhancing usability.
- Introduced a memoized function in the StepEditorContent to determine if existing content is present, streamlining the logic for AI content insertion.

merrcury committed:
feat(dashboard): refactor AI suggestions handling and introduce custom hook

- Refactored AI suggestions logic by creating a new custom hook, `useAiSuggestions`, to encapsulate AI-related functionalities and improve code organization.
- Updated the `StepEditorContent` component to utilize the new hook, enhancing readability and maintainability.
- Streamlined the handling of existing content and AI insertion logic, improving user experience during content generation.
- Removed redundant imports and functions, optimizing the component structure.

merrcury committed:
chore(api): clean up unused code and imports in AI module files

- Removed unnecessary blank lines and unused imports across various AI-related files to improve code readability and maintainability.
- Streamlined the structure of the AI module and its components, ensuring a cleaner codebase.

merrcury committed:
refactor(api): update AI controller to exclude from Swagger documentation

- Removed unused imports and the ApiTags decorator from the AiController.
- Added ApiExcludeController to prevent the AI controller from appearing in Swagger documentation, streamlining API visibility.

scopsy committed:
Merge branch 'next' into feat/step-creation-with-llm

scopsy commented on apps/dashboard/src/components/workflow-editor/steps/email/email-body.tsx:
```diff
@@ -335,7 +335,7 @@ export const EmailBody = () => {
         return (
           <Maily
-            key={`${editorKey}-repeat-block-enabled`}
+            key={`${editorKey}-repeat-block-enabled-${field.value?.slice(0, 50) || ''}`}
```
Why was this needed?

merrcury replied:
When I was inserting the content it wasn't remounting properly. It helped me forced a remount when the content changes significantly. Only preview was showing up but editor was coming blank or unchanged.

djabarovgeorge commented on apps/api/src/app/ai/usecases/generate-content/generate-content.usecase.ts:
```diff
@@ -0,0 +1,264 @@
+
+    let prompt = basePrompt + (channelGuidelines[stepType] || '');
+
+    if (context) {
```
did we check whether providing one or two examples of valid content improves the quality of the results?

djabarovgeorge commented on apps/api/src/app/ai/usecases/generate-content/generate-content.usecase.ts:
```diff
@@ -0,0 +1,264 @@
+    const response = result.object as GenerateContentResponseDto;
+
+    // For email, generate HTML from TipTap JSON
+    if (command.stepType === StepTypeEnum.EMAIL && response.content) {
```
not sure I fully understand this part, does this mean that for now we only support generating content as html?”

merrcury replied:
We are supporting both, I made changes in upcoming to be more deterministic

djabarovgeorge commented on apps/api/src/app/ai/usecases/generate-content/generate-content.usecase.ts:
```diff
@@ -0,0 +1,264 @@
+
+    this.logger.info(`Successfully generated ${command.stepType} content`);
+
+    const response = result.object as GenerateContentResponseDto;
```
should we parse the result to make sure that the response is a proper type response.content GeneratedContent? in case that the generation failed to follow the structure, we need to throw an exception, otherwise the client will behave unexpectedly

merrcury replied:
Sure, Let me take care of this

djabarovgeorge commented on apps/api/src/app/ai/ai.controller.ts:
```diff
@@ -0,0 +1,40 @@
+
+@ApiCommonResponses()
+@ApiExcludeController()
+@Controller({ path: '/ai', version: '1' })
```
i was thinking, maybe we should reuse the existing routes but add a `generate` parameter, since we might end up with many of these new endpoints.
so it could be something like:
POST /workflows/:workflowId/steps
with a body `{ generate: true }`.
has anyone had a chance to look at how other products structure these APIs?

merrcury replied:
Actually, We dont want to have any ai route to be visible on either sdk or openapi spec, keeping them seperate is a good idea. Also We dont want to use these routes with API Key. So that was the reason we kept them seperare.

LetItRock commented on apps/dashboard/src/components/ai-suggestions/initial-state.tsx:
```diff
@@ -0,0 +1,107 @@
+        </div>
+
+        <div className="flex items-center justify-end gap-2">
+          <button
```
please use the `Button` component everywhere possible with these props:
```
<Button
    variant="secondary"
    className="h-[26px]"
    mode="outline"
    size="2xs"
    trailingIcon={RiSparklingLine}
    onClick={onFeelingLucky}
  >
    I'm feeling lucky
  </Button>
```

merrcury replied:
Done

LetItRock commented on apps/api/src/app/ai/usecases/generate-content/generate-content.usecase.ts:
```diff
@@ -0,0 +1,264 @@
+  content: z.array(mailyNodeSchema).describe('Array of content nodes that make up the email body'),
+});
+
+const emailContentSchema = z.object({
```
we have zod schemas defined here: `libs/application-generic/src/schemas/control/email-control.schema.ts`, please take a look

merrcury replied:
These schemas are serving different purposes and Lets keep them separate:
- Control schemas include configuration fields (skip, layoutId, disableOutputSanitization, etc.)
- AI content schemas focus only on generated content (subject, body, actions)

We should be able to use them on entire workflow creation

LetItRock commented on apps/dashboard/src/hooks/use-generate-ai-content.ts:
```diff
@@ -0,0 +1,96 @@
+  const [messages, setMessages] = useState<AiMessage[]>([]);
+  const [lastResponse, setLastResponse] = useState<GenerateContentResponse | null>(null);
+  const [thinkingStartTime, setThinkingStartTime] = useState<number | null>(null);
+  const [thinkingDuration, setThinkingDuration] = useState<number | null>(null);
```
not used

merrcury replied:
removed

merrcury replied:
This was actually added for chat interface but didn't build that. So removing this

djabarovgeorge commented on apps/api/src/config/env.validators.ts:
```diff
@@ -47,6 +47,8 @@ export const envValidators = {
   NOVU_REGION: str({ default: 'local' }),
   NOVU_SECRET_KEY: str({ default: '' }),
   INTERNAL_SERVICES_API_KEY: str({ default: undefined }),
+  // AI Configuration
```
```suggestion
```

merrcury replied:
Removed

LetItRock commented on apps/api/src/app/ai/ai.controller.ts:
```diff
@@ -0,0 +1,40 @@
+export class AiController {
+  constructor(private generateContentUseCase: GenerateContentUseCase) {}
+
+  @Post('/content/generate')
```
sometimes it works sometimes fails:
```
    err: {
      "type": "NoObjectGeneratedError",
      "message": "No object generated: response did not match schema.: Type validation failed: Value: {\"type\":\"response\",\"properties\":{\"aiMessage\":\"Here's a push notification for your crypto pricing alert. Feel free to customize the placeholders with your actual data.\",\"content\":{\"subject\":\"Crypto Alert: {{payload.cryptoName}} Price Update!\",\"body\":\"{{subscriber.firstName}}, {{payload.cryptoName}} crossed ${{payload.price}}. Check it out now!\"},\"suggestedPayload\":{\"cryptoName\":\"Bitcoin\",\"price\":\"25,000\"}}}.\nError message: [\n  {\n    \"code\": \"invalid_type\",\n    \"expected\": \"string\",\n    \"received\": \"undefined\",\n    \"path\": [\n      \"aiMessage\"\n    ],\n    \"message\": \"Required\"\n  },\n  {\n    \"code\": \"invalid_type\",\n    \"expected\": \"object\",\n    \"received\": \"undefined\",\n    \"path\": [\n      \"content\"\n    ],\n    \"message\": \"Required\"\n  }\n]: [\n  {\n    \"code\": \"invalid_type\",\n    \"expected\": \"string\",\n    \"received\": \"undefined\",\n    \"path\": [\n      \"aiMessage\"\n    ],\n    \"message\": \"Required\"\n  },\n  {\n    \"code\": \"invalid_type\",\n    \"expected\": \"object\",\n    \"received\": \"undefined\",\n    \"path\": [\n      \"content\"\n    ],\n    \"message\": \"Required\"\n  }\n]",
      "stack":
          AI_NoObjectGeneratedError: No object generated: response did not match schema.
              at parseAndValidateObjectResult (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/generate-object/parse-and-validate-object-result.ts:53:11)
              at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
              at async parseAndValidateObjectResultWithRepair (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/generate-object/parse-and-validate-object-result.ts:88:12)
              at async fn (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/generate-object/generate-object.ts:421:24)
              at async <anonymous> (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/telemetry/record-span.ts:18:22)
              at async generateObject (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/generate-object/generate-object.ts:261:12)
          caused by: AI_TypeValidationError: Type validation failed: Value: {"type":"response","properties":{"aiMessage":"Here's a push notification for your crypto pricing alert. Feel free to customize the placeholders with your actual data.","content":{"subject":"Crypto Alert: {{payload.cryptoName}} Price Update!","body":"{{subscriber.firstName}}, {{payload.cryptoName}} crossed ${{payload.price}}. Check it out now!"},"suggestedPayload":{"cryptoName":"Bitcoin","price":"25,000"}}}.
          Error message: [
            {
              "code": "invalid_type",
              "expected": "string",
              "received": "undefined",
              "path": [
                "aiMessage"
              ],
              "message": "Required"
            },
            {
              "code": "invalid_type",
              "expected": "object",
              "received": "undefined",
              "path": [
                "content"
              ],
              "message": "Required"
            }
          ]
              at Function.wrap (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/@ai-sdk+provider@2.0.0/node_modules/@ai-sdk/provider/src/errors/type-validation-error.ts:49:9)
              at safeValidateTypes (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/@ai-sdk+provider-utils@3.0.18_zod@3.25.20/node_modules/@ai-sdk/provider-utils/src/validate-types.ts:71:34)
              at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
              at async parseAndValidateObjectResult (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/generate-object/parse-and-validate-object-result.ts:43:28)
              at async parseAndValidateObjectResultWithRepair (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/generate-object/parse-and-validate-object-result.ts:88:12)
              at async fn (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/generate-object/generate-object.ts:421:24)
              at async <anonymous> (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/telemetry/record-span.ts:18:22)
              at async generateObject (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/generate-object/generate-object.ts:261:12)
          caused by: ZodError: [
            {
              "code": "invalid_type",
              "expected": "string",
              "received": "undefined",
              "path": [
                "aiMessage"
              ],
              "message": "Required"
            },
            {
              "code": "invalid_type",
              "expected": "object",
              "received": "undefined",
              "path": [
                "content"
              ],
              "message": "Required"
            }
          ]
              at get error [as error] (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/zod@3.25.20/node_modules/zod/dist/cjs/v3/types.js:57:31)
              at Object.validate (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/@ai-sdk+provider-utils@3.0.18_zod@3.25.20/node_modules/@ai-sdk/provider-utils/src/zod-schema.ts:33:45)
              at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
              at async safeValidateTypes (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/@ai-sdk+provider-utils@3.0.18_zod@3.25.20/node_modules/@ai-sdk/provider-utils/src/validate-types.ts:63:20)
              at async parseAndValidateObjectResult (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/generate-object/parse-and-validate-object-result.ts:43:28)
              at async parseAndValidateObjectResultWithRepair (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/generate-object/parse-and-validate-object-result.ts:88:12)
              at async fn (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/generate-object/generate-object.ts:421:24)
              at async <anonymous> (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/telemetry/record-span.ts:18:22)
              at async generateObject (/Users/pavlotymchuk/projects/js/novu-2/node_modules/.pnpm/ai@5.0.108_zod@3.25.20/node_modules/ai/src/generate-object/generate-object.ts:261:12)
      "name": "AI_NoObjectGeneratedError",
      "text": "{\"type\":\"response\",\"properties\":{\"aiMessage\":\"Here's a push notification for your crypto pricing alert. Feel free to customize the placeholders with your actual data.\",\"content\":{\"subject\":\"Crypto Alert: {{payload.cryptoName}} Price Update!\",\"body\":\"{{subscriber.firstName}}, {{payload.cryptoName}} crossed ${{payload.price}}. Check it out now!\"},\"suggestedPayload\":{\"cryptoName\":\"Bitcoin\",\"price\":\"25,000\"}}}",
```

merrcury replied:
When LLM doesn't follow instruction and try to goes out of instructions. I have improved prompt and parsing for these cases and also if that will not work either we have added retry. These things are going to be reinforcement learning for us also. But for now this is resolved

LetItRock commented on apps/dashboard/src/components/ai-suggestions/ai-suggestions-dialog.tsx:
```diff
@@ -0,0 +1,170 @@
+      onInsert(lastResponse.content, lastResponse.suggestedPayload);
+      handleClose();
+    }
+  }, [lastResponse, onInsert, handleClose]);
```
The insert action should force storing the workflow, please take a look at:
```
const { saveForm } = useSaveForm();
```

merrcury replied:
Done 👍🏼

LetItRock commented on apps/dashboard/src/components/ai-suggestions/ai-suggestions-dialog.tsx:
```diff
@@ -0,0 +1,170 @@
+          ) : isLoading ? (
+            <LoadingState stepType={stepType} />
+          ) : (
+            <ContentPreview
+              stepType={stepType}
+              content={lastResponse?.content}
+              previewPayload={parsedPayload}
+              suggestedPayload={lastResponse?.suggestedPayload}
+            />
```
I think that we should not be parsing the variables in the Dashboard, but instead use the `/workflows/${workflowSlug}/step/${stepSlug}/preview` endpoint to do the job. 

The ai endpoint response should be returning:
- step control values
- preview payload (based on the existing payload and the generated one)

LetItRock replied:
The `usePreviewStep` hook can be useful to make a call.

LetItRock replied:
this is the issue:
<img width="1218" height="783" alt="Image" src="https://github.com/user-attachments/assets/6249dd1b-797c-4c85-b9bc-ecc9e38d25c8" />

LetItRock commented on apps/dashboard/src/components/workflow-editor/steps/use-ai-suggestions.tsx:
```diff
@@ -0,0 +1,235 @@
+                form.setValue('body', bodyStr);
+                form.trigger(['subject', 'body']);
+              });
+            });
```
`requestAnimationFrame` is not needed, just update the form values and trigger workflow form save, check this comment: https://github.com/novuhq/novu/pull/9621/files#r2602780193

merrcury replied:
Done ✅

LetItRock commented on apps/dashboard/src/components/workflow-editor/steps/use-ai-suggestions.tsx:
```diff
@@ -0,0 +1,235 @@
+      }
+
+      // Merge suggested payload into preview sandbox
+      try {
```
Something is not working here:

https://github.com/user-attachments/assets/b3d5c61e-bf5c-439b-b9d8-2952958471db

merrcury replied:
This is fixed

LetItRock commented on apps/dashboard/src/components/workflow-editor/steps/use-ai-suggestions.tsx:
```diff
@@ -0,0 +1,235 @@
+    }
+  }, [form, step.type]);
+
+  const handleAiInsert = useCallback(
```
it should also work with the HTML editor:

https://github.com/user-attachments/assets/34a49714-628b-4036-9cfa-187662c536c7

merrcury replied:
Done ✔️

merrcury committed:
feat(api): introduce MessageRole enum for message sender roles in generate-content-request DTO

- Updated the MessageDto to use the new MessageRole enum for defining message sender roles, enhancing type safety and clarity in the codebase.

merrcury committed:
fix(ai-suggestions): prevent "I'm feeling lucky" button from being clickable when no suggestions are available

- Updated the button to be disabled when there are no suggestions, improving user experience and preventing unnecessary actions.
- Removed unused error handling in the AI content generation hook for cleaner code.

merrcury committed:
fix(ai-suggestions): parse action labels in ContentPreview for improved rendering

- Updated the primary and secondary action labels in the ContentPreview component to use the parse function, ensuring proper rendering of dynamic content.

merrcury committed:
fix(ai-suggestions): enhance Enter key handling in InitialState component

- Updated the Enter key event handler to check if the prompt is not empty before submitting, preventing empty submissions and improving user experience.

merrcury committed:
refactor(ai-suggestions): optimize content existence checks using form watchers

- Replaced direct form value retrieval with watched values for subject and body, improving performance and reactivity in determining existing content for various step types.
- Cleaned up code for better readability and maintainability.

merrcury committed:
refactor(ai-content): streamline state management in useGenerateAiContent hook

- Consolidated state variables by removing thinking duration and start time, simplifying the mutation lifecycle.
- Enhanced message handling in the onSuccess callback for improved readability and maintainability.

merrcury committed:
refactor(env): remove commented AI configuration section from env.validators.ts for clarity

merrcury committed:
refactor(dashboard): replace button elements with Button component for consistency

- Updated AiSuggestionsDialog, ErrorState, InitialState, and StepEditorLayout components to use the new Button component for improved styling and consistency across the application.
- Enhanced user interface elements by replacing standard button elements with the Button component, ensuring a unified look and feel.

merrcury committed:
refactor(ai-suggestions): enhance UI components for better consistency and usability

- Replaced standard HTML elements with respective component imports (Hint and Textarea) in ContentPreview and InitialState for improved styling and consistency.
- Updated loading state styles to utilize Skeleton components, enhancing the visual feedback during loading processes.
- Adjusted text colors for better readability and alignment with design standards.

merrcury committed:
feat(ai-suggestions): add EmailStatePreview component for email loading and error states

- Introduced a new EmailStatePreview component to handle both loading and error states for email suggestions.
- Updated ErrorState and LoadingState components to utilize EmailStatePreview, enhancing code reusability and consistency in UI presentation.
- Removed deprecated EmailErrorPreview and EmailLoadingPreview components to streamline the codebase.

merrcury committed:
refactor(ai-suggestions): enhance UI components and styling consistency

- Updated AiSuggestionsDialog, ContentPreview, EmailStatePreview, and ErrorState components to improve UI consistency by replacing standard elements with respective component imports.
- Adjusted styles and class names for better alignment with design standards, including the use of Badge and Hint components for enhanced visual feedback.
- Improved text readability by updating color classes across various components.

merrcury committed:
refactor(ai-suggestions): reorganize imports in EmailStatePreview component

- Cleaned up import statements in the EmailStatePreview component for better readability and organization.
- Ensured consistent usage of component imports from the primitives directory, enhancing maintainability.

merrcury committed:
feat(ai-suggestions): integrate form saving and schema updates in AI suggestions

- Added functionality to save the workflow form after inserting AI-generated content.
- Implemented schema updates to handle new properties dynamically, ensuring the schema is saved when new properties are added.
- Refactored the `inferTypeFromValue` function to use `unknown` for better type safety.

merrcury committed:
refactor(ai-suggestions): streamline form value updates in AI suggestions

- Simplified the setting of form values for subject and body by consolidating the logic into a single call, ensuring the form state is marked as dirty and touched.
- Removed unnecessary animation frames, improving performance and code clarity.

merrcury committed:
feat(ai-content): add editorType support for AI content generation

- Introduced editorType property in GenerateContentRequest and GenerateContentResponse to specify the editor being used (HTML or block).
- Updated relevant DTOs and use cases to handle editorType, ensuring proper content generation based on the selected editor.
- Enhanced the getSchemaForStepType method to dynamically select the appropriate schema based on editorType.
- Implemented fallback logic for content generation in case of errors, allowing for a switch between HTML and block editors.
- Updated UI components to accommodate the new editorType functionality, improving the overall user experience.

merrcury committed:
chore(api): modularize content generation DTOs and schemas

- Extracted MessageDto and WorkflowContextDto into separate files for better organization and reusability.
- Updated GenerateContentRequestDto to utilize the new DTOs, enhancing clarity and maintainability.
- Introduced content generation schemas for various notification types, streamlining the structure of generated content.
- Added constants for content generation guidelines, improving the overall content generation process.

merrcury committed:
feat(generate-content): enhance content generation structure and error handling

- Updated constants to specify critical output format for generated content, ensuring a flat JSON response structure.
- Improved response schema to clarify the requirements for aiMessage, content, and suggestedPayload fields, emphasizing their placement at the root level.
- Enhanced error handling in the content generation process with retry logic for schema validation failures, improving robustness and user feedback.

merrcury committed:
feat(api): imrpove error handling for external AI Inputs

merrcury commented:
Closed in favour of ongoing AI project
